src/behavior-considered-undefined.md

```diff
@@ -58,7 +58,19 @@ Please read the [Rustonomicon] before writing unsafe code.
 * Invoking undefined behavior via compiler intrinsics.
 * Executing code compiled with platform features that the current platform
   does not support (see [`target_feature`]), *except* if the platform explicitly documents this to be safe.
-* Calling a function with the wrong call ABI or unwinding from a function with the wrong unwind ABI.
+* Calling a function with the wrong call ABI or unwinding from a function with the wrong unwind ABI
+* Causing an unwind into Rust code from a foreign function that was called via a
+  function declaration or pointer declared with a non-unwinding ABI, such as `"C"`,
+  `"system"`, etc. (For example, this case occurs when such a function written in
+  C++ throws an exception that is uncaught and propagates to Rust.)
+* Calling a Rust `extern` function that unwinds (with `extern "C-unwind"` or
+  another ABI that permits unwinding) from a runtime that does not support
+  unwinding, such as code compiled with GCC or Clang using `-fno-exceptions`
+* Catching a Rust `panic` in non Rust code (for instance `catch (...)` in C++)
+* Catching a non-Rust unwind (such as a C++ exception) with `catch_unwind`
```

Comment:
Why is this repeated here? We don't repeat all the UB one can cause by any library function here.

Reply:
It's a special case of a library function that does something no other function can do, though; the fact that this is undefined can't be inferred from the other undefined behavior listed in the document.

See what you think now that the undwinding UB is in its own section.

Reply:
It's not really different from any of the other language intrinsics that all have their own UB, summarized here as just "Invoking undefined behavior via compiler intrinsics". Why would this intrinsic be treated specially?

Reply:
Because there's no inherent reason why this is clearly UB, and I believe multiple people have already expressed surprise that this is UB. The name implies that it should be able to catch any unwinding.

Reply:
I don't see how adding this here helps with that. We explicitly do not list all language intrinsics and their respective UB here. This is not a guide for beginners. The place to fix that surprise is [these docs](https://doc.rust-lang.org/nightly/std/panic/fn.catch_unwind.html).

Reply:
I agree that the doc comment for `catch_unwind` needs to be updated, by why _shouldn't_ this page also include the information?

Reply:
Why should it do that for this one intrinsic but not all the other intrinsics? It may lead someone to conclude that "oh the page doesn't repeat the UB documented for unchecked_add so somehow that's less UB than the UB in catch_unwind that *is* being repeated".

This page is specifically about language primitive operations that are exposed to the user on stable and that can cause UB. Everything that is just exposed via a library function is not on this page. If you want to change that, we can have that discussion, but we'd certainly not make an exception for *one* library function -- we'd then want to find a systematic way to characterize "library functions that expose a language intrinsic" and list *all* of them.

Reply:
Here's a PR to fix the `catch_unwind` docs: https://github.com/rust-lang/rust/pull/128321

I feel weird about it, but I've taken out the explicit mention of `catch_unwind`.